### ISSUE TITLE
CUDA context creation problem in nnet3 training with "--use-gpu=wait" option

### ISSUE THREAD

**Issue by dogancan** [kaldi-asr/kaldi#1487]
I am not sure if this is a Kaldi issue but I thought someone might have an idea.

First some context. I am trying to tune a few TDNN chain models on a workstation with 2 Maxwell Titan X 12GB cards. The data sets I am working with are fairly small (Babel full language packs with 40-80 hours audio). Initially I set the number of initial and final training jobs to 2 and trained the models with scripts adapted from babel and swbd recipes. While this worked without any issues, I noticed that the models were overtraining, so I tried tuning relu-dim, number of epochs and xent-regularize with one of the language packs to see if I could get a better model. Eventually the best model I got was with a single epoch and xent-regularize=0.25 (WER base model: 45.5% vs best model: 41.4%). To see if the training schedule might have any further effects on the model performance, I also tried training with `--num-jobs-initial=2, --num-jobs-final=8` after setting the GPUs to "default" compute mode to allow the creation of multiple CUDA contexts. I added 2 seconds delay between individual jobs so that earlier jobs would start allocating device memory before a new job is scheduled on the device with the largest amount of free memory. This mostly worked fine, except towards the end when 8 jobs were distributed 5-3 between the two cards. The resulting model had 40.9% WER after 2 epochs and the log probability difference between the train and validation sets was also smaller than before. It seems like the training schedule (number of jobs, learning rate, etc. at each iteration) has an effect on the model performance in this small data scenario. Maybe averaging gradients across a larger number of jobs is beneficial, or the learning rate schedule is somehow tuned for this type of training schedule. 

Now the actual problem. Since large number of jobs seemed to work better for me, I wanted to remove the job delay hack, set GPUs back to "exclusive process" compute mode and take advantage of the `--use-gpu=wait` option while scheduling the training jobs. However, it seems like I am missing something. If I launch multiple training processes with the `--use-gpu=wait` option while GPUs are in  "exclusive process" compute mode, only one process can create a CUDA context on a given GPU card even after that one process completes. My expectation was that other processes would wait for the GPUs to be available and then one by one acquire the GPUs and complete their work. I added a few debug statements to `GetCudaContext` function to see what the problem was. `cudaDeviceSynchronize` call returns "all CUDA-capable devices are busy or unavailable" even after processes running on the GPUs are long gone. Any ideas?

**Comment by dogancan** [comment]
I looked into this issue a bit more and it seems the problem is not limited to `--use-gpu=wait` case. If there are no available GPUs when a training process is started, it fails to acquire the GPU after waiting it to be available. I replicated the problem with `--use-gpu=yes` by first launching 2 training processes and starting a third process while the other 2 were about to complete. Below is the log for the third process.

```
nnet3-chain-train --use-gpu=yes --apply-deriv-weights=False --l2-regularize=5e-05 --leaky-hmm-coefficient=0.1 --read-cache=exp/chain/tdnn7_sp/cache.0 --xent-regularize=0.25 --print-interval=10 --momentum=0.0 --max-param-change=1.41421356237 'nnet3-am-copy --raw=true --learning-rate=0.003 exp/chain/tdnn7_sp/0.mdl - |' exp/chain/tdnn7_sp/den.fst 'ark,bg:nnet3-chain-copy-egs --left-context=15 --right-context=15 --frame-shift=0 ark:exp/chain/tdnn_sp/egs/cegs.3.ark ark:- | nnet3-chain-shuffle-egs --buffer-size=5000 --srand=0 ark:- ark:- | nnet3-chain-merge-egs --minibatch-size=64,32 ark:- ark:- |' exp/chain/tdnn7_sp/1.3.raw
WARNING (nnet3-chain-train[5.1.26~1-4a58a]:SelectGpuId():cu-device.cc:137) Will try again to get a GPU after 20 seconds.
Fri Mar 10 15:30:55 2017
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.26                 Driver Version: 375.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TIT...  Off  | 0000:05:00.0      On |                  N/A |
| 32%   67C    P8    19W / 250W |     57MiB / 12203MiB |      0%   E. Process |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TIT...  Off  | 0000:09:00.0     Off |                  N/A |
| 26%   64C    P2    70W / 250W |      3MiB / 12206MiB |      0%   E. Process |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      1451    G   /usr/lib/xorg/Xorg                              53MiB |
+-----------------------------------------------------------------------------+
LOG (nnet3-chain-train[5.1.26~1-4a58a]:SelectGpuId():cu-device.cc:146) num-gpus=2. Device 0: all CUDA-capable devices are busy or unavailable.  Device 1: all CUDA-capable devices are busy or unavailable.
ERROR (nnet3-chain-train[5.1.26~1-4a58a]:SelectGpuId():cu-device.cc:147) Failed to create CUDA context, no more unused GPUs?

[ Stack-Trace: ]

kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::CuDevice::SelectGpuId(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
main
__libc_start_main
_start
```

I can also replicate the same behavior with a single GPU card and 2 parallel processes by restricting the visible GPUs to one of the two cards. I still think this is likely a CUDA or NVIDIA driver issue. I tried reinstalling CUDA and rebuilding Kaldi but that did not change anything. I find it rather unlikely that the gpu waiting code in `cu-device.cc` always had this issue. Maybe CUDA-8.0 is somehow not compatible with the current code? Can anyone with the resources try to replicate this on their end?

**Comment by lifeiteng** [comment]
Maybe there is a problem when multi-processes request the same GPU which under "exclusive process" mode.

When `--num-jobs-final=8` > the number of GPUs, my solution is:
add `--gpu-id=0|1|2|3...` (add SelectGpuId(int32 gpu_id)  method in cu-device.h ) config let `nnet-*-train` know which GPU to use, control this in training script (e.g --num-jobs-final=8, 4 GPUs, parallel job 1-4, then job 5-8, next iter ).

**Comment by danpovey** [comment]
Dogan, I did verify that the problem exists.
I also suspect it is a CUDA 8.0 issue.
This is actually not the first time this has happened.  A few versions ago,
we had to rewrite the code that handles getting the device because the
behavior of those functions changed.  IIRC neither the behavior before nor
after the change was documented.  So I guess they've changed things under
the hood again.

**Comment by danpovey** [comment]
BTW, one possible solution that it would be nice if you could try is here:

    cudaSetDevice(device);

    cudaError_t e = cudaDeviceSynchronize(); // << CUDA context gets
created here.



instead of checking the error status of cudaDeviceSyncronize(), check it
from cudaSetDevice() instead, and only if that succeeds, just call
cudaDeviceSynchronize() as a kind of double-check, and crash if it fails.
Because if cudaSetDevice() succeeds, we should be OK to use the device.


Dan

**Comment by dogancan** [comment]
I can look into this tomorrow but I don't think cudaSetDevice() works as a way of checking if the current process acquired the context.

**Comment by dogancan** [comment]
TL;DR It seems like there is a bug (or limitation) in the CUDA Runtime API -- not sure if it also affects the CUDA Driver API. I don't think there is anything we can do until CUDA fixes it.

There is a little bit of useful information available in the description of computeMode in the description of [cudaGetDeviceProperties](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g1bf9d625a931d657e08db2b4391170f0):

> computeMode is the compute mode that the device is currently in. Available modes are as follows:
cudaComputeModeDefault: Default mode - Device is not restricted and multiple threads can use cudaSetDevice() with this device.
cudaComputeModeExclusive: Compute-exclusive mode - Only one thread will be able to use cudaSetDevice() with this device.
cudaComputeModeProhibited: Compute-prohibited mode - No threads can use cudaSetDevice() with this device.
cudaComputeModeExclusiveProcess: Compute-exclusive-process mode - Many threads in one process will be able to use cudaSetDevice() with this device.
If cudaSetDevice() is called on an already occupied device with computeMode cudaComputeModeExclusive, cudaErrorDeviceAlreadyInUse will be immediately returned indicating the device cannot be used. When an occupied exclusive mode device is chosen with cudaSetDevice, all subsequent non-device management runtime functions will return cudaErrorDevicesUnavailable.

Unfortunately they don't say what [cudaSetDevice](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g159587909ffa0791bbe4b40187a4c6bb) does if it is called on a device with computeMode cudaComputeModeExclusiveProcess that is already occupied by another process . Now I would expect it to return [cudaErrorDeviceAlreadyInUse](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038459c4ec80247b489ddb9c5577fcc266a) indicating the device cannot be used, but that is not what happens. In fact it returns [cudaSuccess](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038e355f04607d824883b4a50662830d591). They again don't say what subsequent non-device management runtime functions return when an occupied "Exclusive Process" mode device is chosen with cudaSetDevice from a different process. This time they do what I expect them to do and return [cudaErrorDevicesUnavailable](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038a2ab67256308f82a7be0fcd9fb145ad7). The problem is they continue to return cudaErrorDevicesUnavailable even if the device is no longer occupied. I tried calling [cudaDeviceReset](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gef69dd5c6d0206c2b8d099abac61f217) after every failed context creation attempt but that did not change anything. I tried calling cudaFree(0) instead of cudaDeviceSynchronize() but that did not change anything either. I also tried a bunch of other hacks that I can not remember now but nothing helped. It seems like there is nothing we can do at this point other than submitting a bug report to NVIDIA and wait for them to fix it. If anyone reading this can think of anything to work around this issue in the meantime, let me know.

**Comment by danpovey** [comment]
OK, thanks for looking into it.

**Comment by vince62s** [comment]
I am interested by both understanding  '--use-gpu=wait ' is this supposed to avoid setting up a grid set-up by queuing some job batches waiting availability ? and also by the fact that gradient averaging seems to give better results. If this is really the case, couldn't we schedule the jobs differently by batches even when we have only one or two GPUs on the same computer (ie run mode vs grid mode).

**Comment by danpovey** [comment]
I haven't really been using that functionality; Karel wrote it and I don't
know what he was originally using it for.  In principle you could use it
that way but until we can figure out what the problem is with the CUDA
toolkit version 8, it may not be wise to rely on it working.

Dan

**Comment by dogancan** [comment]
Yes, the use case for the `--use-gpu=wait` option is to queue jobs that will wait for availability. The problem is not limited to the `--use-gpu=wait` option though. Even with the `--use-gpu=yes` option, Kaldi programs that fail to create a CUDA context in their first attempts will wait for 20 seconds to try one more time. Independent of whether any GPUs are available at the time or not, that second attempt will fail with cudaErrorDevicesUnavailable. I believe this code used to work before but it doesn't work anymore with recent CUDA toolkits. Even if we schedule the jobs in batches, if we want to use the GPUs in exclusive process mode we still need to somehow ensure that GPUs are not used by other processes when our jobs try to create CUDA contexts. Basically we need a job scheduler like a grid engine :) Regarding gradient averaging giving better results with larger number of jobs, I doubt that is true in general. Even in my case, the improvement might not be statistically significant. It might be that the current learning schedules are better tuned for a larger number of jobs running in parallel. Maybe Dan can comment on these.

**Comment by danpovey** [comment]
More jobs is not always better; it tends to lead to slower optimization
which can help if overtraining is an issue and you have a lot of epochs,
but that's basically a tuning issue.

It would be great if someone could figure out a way to resolve the problem
of selecting a free GPU.

**Comment by jtrmal** [comment]
I didn't manage to follow the whole thread, so apologies if I'm off-topic,
but:

> It would be great if someone could figure out a way to resolve the
problem of selecting a free GPU.

unless there is documented driver-api, perhaps at least nvidia-smi could be
used?

$ nvidia-smi --query-gpu=timestamp,gpu_serial,gpu_uuid,index --format=csv
timestamp, serial, uuid, index
2017/03/12 16:42:42.695, 0324815000345,
GPU-5c3afc2f-4257-841a-8e7d-f0bc70352ad3, 0
2017/03/12 16:42:42.695, 0324815000345,
GPU-f57f24bf-3538-305d-7fc5-699ca3dd9ce5, 1
2017/03/12 16:42:42.696, 0324815000460,
GPU-c78e6c2c-21aa-15b5-7819-d2cbffd22a96, 2
2017/03/12 16:42:42.696, 0324815000460,
GPU-fd3b4bbd-103f-4000-565a-92e7e6db37d4, 3
$ nvidia-smi --query-compute-apps=timestamp,gpu_serial,gpu_uuid,pid,name
--format=csv
timestamp, gpu_serial, gpu_uuid, pid, process_name
2017/03/12 16:42:49.773, 0324815000345,
GPU-5c3afc2f-4257-841a-8e7d-f0bc70352ad3, 6489, nnet-train-frmshuff-ncca
2017/03/12 16:42:49.774, 0324815000345,
GPU-f57f24bf-3538-305d-7fc5-699ca3dd9ce5, 35992, python
2017/03/12 16:42:49.775, 0324815000460,
GPU-c78e6c2c-21aa-15b5-7819-d2cbffd22a96, 46206, python
2017/03/12 16:42:49.776, 0324815000460,
GPU-fd3b4bbd-103f-4000-565a-92e7e6db37d4, 136576, python

from these two outputs, it should be possible to work out which gpu is
probably free?
y.

**Comment by danpovey** [comment]
This approach would not work for GTX GPUs because in that case NVidia
doesn't tell you about the processes.

**Comment by vince62s** [comment]
what about based on just querying the free memory of the device ? (this supposes we know the card type and amount of memory needed for a job, or just a yes/no assumption)

**Comment by jtrmal** [comment]
I think that's exactly how is it done now.
y.

**Comment by danpovey** [comment]
The memory is only relevant in non-exclusive mode.   The problem here is
all about the behavior in exclusive-mode.  Memory doesn't tell you all that
you need because someone could have grabbed a GPU but not be using memory.

I think it might be possible to do it with cudaChooseDevice-- just don't
give it any properties (we'll assume that all visible devices are
suitable).  Presumably it is smart enough to know not to choose an occupied
device in exclusive mode.

**Comment by danpovey** [comment]
BTW, someone here
https://devtalk.nvidia.com/default/topic/415723/choosing-cuda-device-programmatically/
is saying that when using compute exclusive mode, you should just not call
cudaSetDevice-- presumably it will automatically pick a free device when
you call cudaDeviceSynchronize().
Interestingly, this is how the code *originally* worked.  However, after a
certain version of the CUDA toolkit, the original code stopped working, and
that's when we added the iteration over devices and the calls to
cudaSetDevice.
One possibility is to query the version of the CUDA toolkit, and if it's >=
8.0, go back to the original strategy of trying to call
cudaDeviceSynchronize() and checking the error status.

**Comment by dogancan** [comment]
It seems like this problem is not related to cudaSetDevice or cudaDeviceSynchronize or any other CUDA API call in particular. I also looked into using the CUDA Driver API but that does not work either. I am starting to think this is not a bug but a feature :) Are we sure this mechanism worked before with exclusive compute mode?

**Comment by danpovey** [comment]
Yes I'm sure it worked before.
If it's a feature then it's a new feature.  As I say, it's not the first
time they have messed with this stuff without releasing any documentation
about what was changed or why.
The ability to 'wait' for a GPU (and the fact that it retry after 20
seconds if you didn't ask it to wait) are not things that any of the
checked-in Kaldi scripts rely on, although the wait-for-20-secs thing does
improve robustness.
One possibility is to just remove or temporarily disable the "wait" option,
so that people don't try to use it and then get confused.

**Comment by dogancan** [comment]
Ok. I have narrowed down the problem to second process trying to create a CUDA context before the first process destroys its CUDA context. If a process tries to create a CUDA context on a device in exclusive process mode and fails because another process currently owns a context on that device, it is doomed. There is no recourse. Process lifetimes are not relevant. Say if the first process destroys its context and continues to do other things, second process can come along and create a CUDA context as long as that context creation call happens after the first process destroys its context.

**Comment by dogancan** [comment]
I forgot to mention that calling cudaSetDevice or any other cuda function that doesn't attempt to create a context from the second process before the first process destroys its context is OK. Only those functions that attempt to create a context, such as cudaDeviceSynchronize and cudaFree, need to happen after the first process destroys its context.

**Comment by danpovey** [comment]
But I assume there is no way to find out whether the GPU is blocked except
for calling cudaDeficeSynchronize and the like?

It occurred to me that possibly NVidia is deliberately making this
difficult for us because if we were able to do this, it would be possible
to easily figure out whether GPUs are being used even on low-end GPUs like
GTX; and this would enable easier workarounds for at least some of the
functionality of nvidia-smi that is disabled for low-end GPUs.

**Comment by dogancan** [comment]
Yes, unfortunately that seems to be the case.

**Comment by vesis84** [comment]
Hi Dan,
it is not true that I have created the the '--use-gpu=wait' function.

It was created by your students, as can be seen in the log:

7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 133)   if 
(use_gpu != "wait") {
6548565445 (Dan Povey         2015-07-16 05:00:44 +0000 134)     if 
(!got_context) {
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 135)       // So 
far no we don't have context, sleep a bit and retry.
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 136)       int32 
sec_sleep = (use_gpu == "yes" ? 20 : 2);
3dd99870b6 (Kirill Katsnelson 2015-07-11 01:49:16 +0000 137)       
KALDI_WARN << "Will try again to get a GPU after " << sec_sleep
3e6dedb69a (Dan Povey         2015-12-02 15:42:42 -0500 
138)                  << " seconds.";
3dd99870b6 (Kirill Katsnelson 2015-07-11 01:49:16 +0000 139)       
Sleep(sec_sleep);
9f44f1e407 (Daniel Povey      2015-12-10 16:00:10 -0500 140)       if 
(!GetCudaContext(num_gpus, &debug_str)) {
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 141)         if 
(use_gpu == "yes") {
fb6eea05b0 (Minhua Wu         2015-12-07 22:22:47 -0500 142)           {
fb6eea05b0 (Minhua Wu         2015-12-07 22:22:47 -0500 143)             
Input input;
fb6eea05b0 (Minhua Wu         2015-12-07 22:22:47 -0500 144)             
input.Open("nvidia-smi 1>&2 |");
fb6eea05b0 (Minhua Wu         2015-12-07 22:22:47 -0500 145)           }
9f44f1e407 (Daniel Povey      2015-12-10 16:00:10 -0500 146)           
KALDI_LOG << debug_str;
9f44f1e407 (Daniel Povey      2015-12-10 16:00:10 -0500 147)           
KALDI_ERR << "Failed to create CUDA context, no more unused GPUs? ";
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 148)         }
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 149)         if 
(use_gpu == "optional") {
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 150)           
KALDI_WARN << "Running on CPU!!! No more unused CUDA GPUs?";
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 151)           
return;
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 152)         }
c3a5fa2187 (Karel Vesely      2013-11-21 21:24:54 +0000 153)       }
c3a5fa2187 (Karel Vesely      2013-11-21 21:24:54 +0000 154)     }
7b8e847d4f (Vimal Manohar     2015-01-14 08:28:11 +0000 155)   } else {

Originally there were three options that we agreed on back in Nov2013: 
--use-gpu={yes,no,optional}
I am sorry for complaining, but I am really not happy when a wrong 
information about me is spreaded to the community.
It really touches me personally...

Karel.

**Comment by danpovey** [comment]
Oh, OK.  I wasn't complaining about the option- I see that it can be useful
sometimes- I just misremembered why it was created and by whom.

**Comment by vince62s** [comment]
on my end I am very excited by this option ... so if used in non-exclusive mode, is it working as expected based the free-mem availability ? I had the impression that current implementation was just picking the GPU with the most available free memory. But if there was a waiting condition based on eg. 90% of free memory, wouldn't that work (just in non exclusive mode) ?

**Comment by danpovey** [comment]
No it wouldn't work, because what if two jobs start at the same time?

**Comment by vesis84** [comment]
Hi,
I just went through the thread. Okay, I see the that '--use-gpu=wait' can be helpful,
we should ask NVidia guys, how to correctly implement the scenario:
- there's process exclusive mode,
- the GPU is taken by another process,
- how should our process query if the GPU is still occupied or already had been freed,
K.

**Comment by vesis84** [comment]
Here is some info from Jeremy: 

> At first glance this seems like a bug – once there is no longer a process on the GPU it shouldn’t be reporting that there is.
> If you could provide a small reproducing example I can have a play around with it and confirm I see the same issue. In the meantime, it’s probably best to file a bug from the CUDA registered developer website. If you could then forward me the bug ID it would be great.

**Comment by danpovey** [comment]
That's useful.  Do you have time to file a bug as he says, or maybe you can
ask one of the students that you supervise to help?  If not I can try to
find someone.

**Comment by dogancan** [comment]
I have a minimal example that we can use. I can also do the bug report.

**Comment by danpovey** [comment]
That's great-- thanks!  I'll include you in the conversation with the
nvidia guys, Karel cc'd me.

**Comment by JoshLondon** [comment]
I'm having this same problem, tried fiddling with the code as well to no avail and then came across this useful thread. My problem is I can't use my 3 x 980ti GPUs, it only uses 1 (In Default) and then in Exclusive mode on training 8 jobs, the 4th stops the training with a problem it can't find a GPU. I normally just make sure the num-jobs == num-gpucards to get around this problem. But for the tdnn/nnet3 it's not like that. So any update on this problem would be amazing please. Thanks.

**Comment by danpovey** [comment]
Regarding the number of GPUs, the nnet3 tools do use more than one GPU, but
that is configurable, as --num-jobs-initial and --num-jobs-final normally.
It will affect the training a bit if you change them, but not a vast amount.
Your choice is either to reduce --num-jobs-final to the number of physical
GPUs, or to install GridEngine.

Unfortunately NVidia's latest drivers (8.0) have broken the --wait option
and there does not seem to be any way to work around it until they fix
things on their side.  So that is not an option.

**Comment by vince62s** [comment]
With these new gpu like gtx1080ti there is really a nice gain in running 2 or more jobs on the same gpu.
So when you 2 or more gpu in the same machine, it would be nice to be able to several jobs on several gpu.... just trying :)

**Comment by dogancan** [comment]
I recently got a notification from nvidia that this bug was fixed. I have no idea when the fix will hit the public CUDA release though.

**Comment by danpovey** [comment]
Hm.  That would require some way of figuring out which GPUs already have
jobs running and assigning jobs in such a way that (say) no GPU has more
than 2 jobs.  You'd have to run GPUs in non-exclusive mode and control it
at the queueing level by setting CUDA_VISIBLE_DEVICES to the appropriate
value.

An alternative, though, would be to have Kaldi figure it out when
non-exclusive mode is detected, by cycling through the GPUs and querying
NVidia functions to figure out how many processes are running on each GPU,
in the same way that it currently figures out how much memory is used on
each-- and choosing the GPU with the fewest number of processes running.
That would be a more sensible rule than querying the memory, because when a
process has just started using a GPU, it may not be using much memory but
it will eventually.  @vince62s, it would be great if you could  figure out
what code changes would be required in Kaldi to do this.
I don't know whether NVidia provides a way to say which processes are using
a GPU, or how many there are, when the GPU is in non-exclusive mode.  And I
don't know whether such a list would include processes that are not really
using the GPU, such as the NVidia persistence daemon.  Could you please
find out about this?

**Comment by entn-at** [comment]
@vince62s I think that's possible as is right now, as long as you make sure that GPU memory isn't a problem. I ran several jobs per GPU (in normal model, "nvidia-smi -c 0" IIRC) by simply increasing the number of slots per machine in the SGE GPU queue. Those jobs were for DNN posterior computation in the SRE10 recipe, so things might be different for other purposes. (CUDA Version 8.0.61, driver version 375.39, Ubuntu 16.04)

**Comment by danpovey** [comment]
The issue I want to be addressed is where there are more than one GPU per
node-- to ensure that the allocation of jobs to GPUs is balanced.
Currently such balancing may be inconsistent
However, if the GPUs are identical (have the same amount of memory) and you
don't attempt to run more than 2 jobs per GPU, it will currently already be
balanced.  Since this is the common case, it may not even be necessary to
make code changes to have this work.

**Comment by vince62s** [comment]
you are right.
however I looked a bit and I think there are some ways to retrieve infos.
On kaldi side, I think the only part linked to this is src/cudamatrix/cu-device.cc
On Cuda side, in fact we need to use NVML.
nvidia-smi actually uses NVML.

For instance, we can work with http://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html
and more specifically: nvmlDeviceGetComputeRunningProcesses

For someone interested in deeper ifo, there is a lot lot to look at.

**Comment by vesis84** [comment]
Perhaps, the ultimate solution would be to use a combination 'process-exclusive' allocation of GPUs together with 'multit-hreading' ? :) Like this, the process gets the whole GPU, and the threads running parallel tasks... This is also beneficial for cache hit-ratio, as all the concurrent tasks are within single GPU :D What do you think? :)

**Comment by vesis84** [comment]
I used to have a multi-threaded scheme implemented in TNet :) There was a 'Platform' class, which contained the FIFO's with features and targets and several copies of NNets with shared weights...

The package with the abandoned TNet project is available on this site : 
http://speech.fit.vutbr.cz/software/neural-network-trainer-tnet
(I could post it to my git account, if it appears to be helpful...)
K.

**Comment by vesis84** [comment]
But, but, but, ...,  before thinking of trying that practically... 

@vince62s : Could you please verify that you use the latest version of kaldi ? 
And also that the verbosity level of NN training tools is set to 0 ?

We have recently changed the code to have less calls to `cuda_device_synchronize()`, which accelerated the BLSTM training considerably (>2x) ... It could be the case, that depending on the 'Kaldi' version, the speed tests may lead to different conclusions, so please make sure you have the latest version compiled :D
K.

**Comment by vince62s** [comment]
Yes, I have the very last code, but okay, I will benchmark exactly the gain when I get some gpus available.

**Comment by vesis84** [comment]
In the past (2010) the total run-time of 2 concurrent GPU jobs was longer than running them one after another... I am very curious about the 'new' run-times :P

**Comment by vince62s** [comment]
Karel, 7 years later ....
njobs	 1	   2         3
iterations	 18	  9	       6
duration	09:17 05:34  04:42

2 comments: 
-this was done on a GTX 1080 (not the Ti)
-the increase of 2 vs 1 job is much higher than before the May 13 commit (IIRC in december the gain was about 20%=

Dan,
I still have an issue. The machine on which we did that has 2 GTX.
Whatever we did (in these runs) all jobs went on the same card, the other having the display taking 300MB on it.
Running 4 jobs blew the memory since one jobs takes 2.4 GB and the card is 8GB.

I am afraid the "memory load" balancing is not working properly.
To anticipate your question: CUDA_VISIBLE_DEVICES was not set.

**Comment by dogancan** [comment]
Hi Vincent

The problem with "memory load" balancing is that the training script launches all jobs one after the other without any delay. Since it takes a few seconds for the job to start consuming memory on the selected GPU, all jobs end up going to the GPU with minimum memory use at the time jobs a re launched. To work around this, you can add a sleep call at the end of the loop launching these jobs.

```
diff --git a/egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py b/egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py
index f28aa89..f016a79 100644
--- a/egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py
+++ b/egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py
@@ -12,6 +12,7 @@ import logging
 import math
 import os
 import sys
+import time

 import libs.common as common_lib
 import libs.nnet3.train.common as common_train_lib
@@ -204,6 +205,7 @@ def train_new_models(dir, iter, srand, num_jobs,
             wait=False)

         processes.append(process_handle)
+        time.sleep(3)

     all_success = True
     for process in processes:
```

**Comment by vesis84** [comment]
@vince62s Thank you for the 'hard data', this is indeed a big step forward ;) I am thinking where to try the multi-threaded processing with 1 GPU first. Perhaps in sMBR training? (seems like a good fit, and it seems this will be a good opprtunity to start using cuda_streams) :) K.

**Comment by vince62s** [comment]
I see where you are coming from, but honestly the use case would be very limited. If you want to fire 12 jobs at the same time you will need 2 or more GPUs anyway, so the way the architecture is today seems fine to me. But intellectually speaking, yes I agree, would be nice.

**Comment by danpovey** [comment]
I had assumed that he was talking about training jobs... turns out he is
talking about the i-vector extraction based on DNNs, in
egs/sre10/v2/sid/extract_ivectors_dnn.sh.  This is a very atypical case
because that command is part of a pipe where other nontrivial stuff is
going on.  It could be that it's not even limited by GPU.

So let's not jump to any conclusions based on his findings.  It could be
that for training, Karel's original observations still hold.

**Comment by vesis84** [comment]
I was thinking of re-implementing 'nnet1' sMBR so it is multi-threaded, and the model averaging would be done asynchronously inside the single training binary using one GPU. If it decreases the training time 2x by using 3 threads, this speed-up would be already very interesting, while the complexity of the code would not be terrific ;) K.

**Comment by danpovey** [comment]
Something that I was considering using for the RNNLM stuff I'm working on
is "multi-stream" processing with GPUs, where different threads have
different streams so their calls to kernels don't have to be serialized.

To make this work automatically (without having to explicity pass around
the streams), apparently you have to pass in the option  --default-stream
per-thread to nvcc (see:
https://devblogs.nvidia.com/parallelforall/gpu-pro-tip-cuda-7-streams-simplify-concurrency/
).

It would be helpful to me if you could use that as a test case as to
whether this works.

**Comment by vesis84** [comment]
Yea, this seems as a good fit ;) (there's usually not too much training data) 

Also, I am thinking of the possible 'catches', one of them is that some of the GPU operations involve an implicit `cuda_device_synchronize()` like the I/O operations for host-to-GPU memory transfers... (while the I/O can be called as a 'stream' specific operation, so it does not block the other threads)

**Comment by vesis84** [comment]
I seems like all we need to do is to create a multi-threaded app, and make sure we replace the cudaDeviceSynchronize() by cudaStreamSynchronized(), since CUDA 7.0 the threads implicitly define the streams :) It's actually a cool idea to have it like that ;)

**Comment by danpovey** [comment]
Karel, I'm pretty sure after reading some stuff about that, that any memory
transfers that involve an implicit synchronize, only synchronize within the
current stream.  Of course, if that's the "legacy default" stream it would
synchronize all streams, but if we compile with the --default-stream
per-thread option, it would just be the current thread's default stream.

 I've seen things saying clearly that memory transfers can take place in
parallel with computation in other threads  that's going on on the GPU.

From what I can see, the appropriate synchronization command would be

cudaStreamSynchronize(cudaStreamPerThread).

**Comment by vesis84** [comment]
Yes, I agree with everything. I think we can switch to '–default-stream per-thread' and  'cudaStreamSynchronize()' already now, when we are using the 'default' thread. Functionally, this will do nothing, and we'll get the performance boost, once we start using multi-threading... :) I am thinking of who might look into it, I'll be a bit busy these days... (ASRU paper, vacation, thesis, work for SoapBox, summer school of 'baroque' style choral singing ;) )

**Comment by vince62s** [comment]
@dogancan I just saw a patch release from nvidia for 8.0.61

**Comment by kkm000** [comment]
@vince62s, @dogancan: But I do not think it does the trick :(

Apparently, the fix is not yet released.

**Comment by vince62s** [comment]
I would like to comment again on this specific issue.
I am running experiments on a 4 GPU machine.
I don't want to set exclusive mode since each GPU can fit 2 jobs at the same time without losing too much computing capability.
As @dogancan said, I ha d to tweak a bit because there is a delay between the time a job is submitted to cuda and the time cuda actually allocate the memory. I had to set up to 8 seconds the time.sleep() in acoustic_model.py to avoid submitting a job to the same "free" GPU twice in a row.

In the end, on a run which has 911 iterations (from 2 to 6 jobs, ie 3648 jobs) I am losing 6 hours of wall clock time. too bad on an actual training time of 13 hours ....

**Comment by danpovey** [comment]
You could make a local script-level change to, for each job, run it as
CUDA_VISIBLE_DEVICES=n [rest of command line]
where n is (job-id % 4).
that would make your hack unnecessary.

**Comment by vince62s** [comment]
much better hack, thanks. still a hack :)

**Comment by AntonOkhotnikov** [comment]
Hi there, are there any updates on the reported issue?

I'm using 2 Tesla GPUs and run them in a Default compute mode (`nvidia-smi -c 0`) because it looks more efficient rather than having 1 process per GPU with Exclusive process mode. Depending on the process, I can accommodate more than 1 process per GPU (Teslas are 16 GB each).

But problem is that all the processes are allocated to 2 GPUs at the same time, and since there is no queuing I'm facing the out of GPU memory error and some jobs fail. In particular, I'm talking about the `steps/nnet/decode/sh --cmd "run.pl" --use-gpu yes --nj 16` command.

Do you have any solution how to run it efficiently using the conditions mentioned above? A lot of small processes with "--use-gpu wait" option and Exclusive mode gives the result, but this seems very inefficient. 

NVIDIA CUDA 10.0

**Comment by danpovey** [comment]
You shouldn't look at it only from the point of view of memory but also CPU utilization.  The gains to be had from sharing would not be that large.  Short answer: there isn't a way to do it the way you want to do it.

Closing the original issue because I don't think it is likely still a problem (initialization code has changed since then).
